### ISSUE TITLE
torch.cat in for-loop failed while parsing

### ISSUE THREAD

**Issue by lygbuaa** [onnx/onnx-tensorrt#852]
[trt_gather_demo.zip](https://github.com/onnx/onnx-tensorrt/files/8924350/trt_gather_demo.zip)
## Description
I am doing torch.cat in a for loop in my model, the code looks like this:

```
import torch

class TrtDemo(torch.nn.Module):
    def __init__(self):
        super(TrtDemo, self).__init__()

    def forward(self, logits:torch.Tensor, indices:torch.Tensor):
        N = logits.shape[0]
        results = torch.zeros(size=[1, 1, 28, 28], dtype=torch.float)
        for idx in range(N):
            dim0 = torch.tensor([idx], dtype=torch.int64)
            dim1 = torch.tensor(indices[idx], dtype=torch.int64)
            # this line is equivalent to "logits_i = logits[idx, indices[idx], :, :].unsqueeze(0).unsqueeze(0)"
            logits_i = torch.index_select(torch.index_select(logits, 0, dim0), 1, dim1)
            results = torch.cat([results, logits_i], dim=0)
            print("idx: {}, logits_i: {}, results: {}".format(idx, logits_i.shape, results.shape))
        return results

if __name__ == "__main__":
    model = TrtDemo()
    model.eval()
    model_jit = torch.jit.script(model)
    logits = torch.rand(size=[10, 4, 28, 28], dtype=torch.float)
    indices = torch.rand([10], dtype=torch.float).type(torch.int64)  
    torch.onnx.export(
    model=model_jit, 
    args=(logits, indices), 
    f="./trt_demo.onnx", 
    opset_version=13, verbose=True, do_constant_folding=True)
```

First export the model to onnx, then  run  `onnx2trt trt_demo.onnx -o demo.trt`  to get the trt engine. It's quite straightforward.
The problem is , when I was using the line `logits_i = logits[idx, indices[idx], :, :].unsqueeze(0).unsqueeze(0)`, I ran into error:
```
[2022-06-15 03:48:58   ERROR] [graphShapeAnalyzer.cpp::throwIfError::1306] Error Code 2: Internal Error (no further information)
While parsing node number 3 [Loop -> "5"]:
ERROR: ModelImporter.cpp:179 In function parseGraph:
[6] Invalid Node - Concat_11
[graphShapeAnalyzer.cpp::throwIfError::1306] Error Code 2: Internal Error (no further information)
```

I suppose tensor shape after slicing operation could be a problem, so I change the line to  `logits_i = torch.index_select(torch.index_select(logits, 0, dim0), 1, dim1)`,  the I ran into another error:

```
[2022-06-15 04:07:57   ERROR] [graphShapeAnalyzer.cpp::throwIfError::1306] Error Code 4: Internal Error ((Unnamed Layer* 9) [Recurrence]: inputs to IRecurrenceLayer have different dimensions. First input has dimensions [1,1,28,28] and second input has dimensions [2,1,28,28].)
While parsing node number 3 [Loop -> "5"]:
ERROR: ModelImporter.cpp:179 In function parseGraph:
[6] Invalid Node - Loop_3
[graphShapeAnalyzer.cpp::throwIfError::1306] Error Code 4: Internal Error ((Unnamed Layer* 9) [Recurrence]: inputs to IRecurrenceLayer have different dimensions. First input has dimensions [1,1,28,28] and second input has dimensions [2,1,28,28].
```

**So, any suggestions, I just want to parse torch.cat in a for-loop .**

## Environment

**TensorRT Version**: 8.0.3
**ONNX-TensorRT Version / Branch**: main branch
**GPU Type**:  RTX A6000
**Nvidia Driver Version**: 470.103.01
**CUDA Version**: 11.4
**CUDNN Version**: 8.2.4
**Operating System + Version**: ubuntu 20.04
**Python Version (if applicable)**: 3.8.12
**TensorFlow + TF2ONNX Version (if applicable)**: 
**PyTorch Version (if applicable)**: 1.10.1
**Baremetal or Container (if container which image + tag)**:  nvcr.io/nvidia/pytorch:21.10-py3

## Steps To Reproduce
you can reproduce easily using my code sample.

**Comment by nvpohanh** [comment]
@lygbuaa Using loops for this looks like an overkill to me. I think there is a way to implement this without using a for-loop, like using torch.gather(). Will get back to you if I have good idea.

**Comment by lygbuaa** [comment]
@nvpohanh Thanks for your quick reply! 
The problem is, this module lies in a detection head, so the tensor shape is inconstant.
could your share a example showing torch.gather() in this kind of inconstant situation?

**Comment by lygbuaa** [comment]
@nvpohanh  Hi, I managed to refactor the code, it's quite straightforward, like this:
```
    def forward(self, logits:torch.Tensor, indices:torch.Tensor):
        index = indices.view(-1, 1, 1, 1).expand(logits.shape)
        # results = torch.gather(logits, 1, index)
        results = logits.gather(1, index)
        return results
```
torch.onnx.export() make no trouble, but when I run `onnx2trt trt_demo.onnx -o demo.trt`, I got `Segmentation fault (core dumped)` when parsing node GatherElements, I can't find the core file (my ulimit -c is unlimited). so what's the problem could be?
seems there's a similar issue posted https://github.com/NVIDIA/TensorRT/issues/1692, is it solved?

**Comment by nvpohanh** [comment]
@lygbuaa That would be a bug we want to fix. Could you first try TRT 8.2 or 8.4 and see if the issue has been fixed? We have improved our support for GatherElements op recently.

**Comment by lygbuaa** [comment]
> 

Nice to hear that, I will try it later. THX

**Comment by lygbuaa** [comment]
> @lygbuaa That would be a bug we want to fix. Could you first try TRT 8.2 or 8.4 and see if the issue has been fixed? We have improved our support for GatherElements op recently.

@nvpohanh Unfortunately, I rebuild onnx-tensorrt with TensorRT 8.4.1(main branch), still got `Segmentation fault (core dumped)`, GatherElements parsing bug still not solved.

**Comment by nvpohanh** [comment]
I see. could you share the ONNX file you exported? We will use it to fix this issue. Thanks

**Comment by lygbuaa** [comment]
> I see. could you share the ONNX file you exported? We will use it to fix this issue. Thanks

@nvpohanh  onnx uploaded to https://github.com/onnx/onnx-tensorrt/files/8924350/trt_gather_demo.zip

**Comment by nvpohanh** [comment]
@lygbuaa I tried this model using trtexec and it worked for me: `trtexec --onnx=build/trt_gather_demo.onnx --verbose --saveEngine=my_engine.plan`. So it may be a bug in onnx2trt tool instead. Do you have to use onnx2trt? Or is it good to build the engine with `trtexec`?

**Comment by lygbuaa** [comment]
@nvpohanh I tried `trtexec` but again got `Segmentation fault (core dumped)`, my trtexec version  is `TensorRT.trtexec [TensorRT v8003]`, so which version are u on?

**Comment by nvpohanh** [comment]
@lygbuaa I am using TRT 8.4.1 (8.4 GA)

**Comment by lygbuaa** [comment]
> @lygbuaa I am using TRT 8.4.1 (8.4 GA)

OK, maybe this is fixed by 8.4.1, I will upgrade my TRT, thanks.
By the way, could you please make a comment on the other issue: https://github.com/onnx/onnx-tensorrt/issues/853
